### PR TITLE
DEPR: Deprecate pandas/io/date_converters.py

### LINKED ISSUE #24518: DEPR: deprecate pandas/io/date_converters.py
This file contains one function that is used elsewhere in the code base (that can be moved), and the rest is unused and not advertised in the api.

Raising this issue to gauge whether we should outright remove it or deprecate this file since its technically public in `io`

### PR DESCRIPTION
**Pull request by avinashpancham**
- [x] closes #24518
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/tools/datetimes.py [hunk 1]
```diff
@@ -59,7 +60,7 @@
 # ---------------------------------------------------------------------
 # types used in annotations
 
-ArrayConvertible = Union[List, Tuple, ArrayLike, "Series"]
+ArrayConvertible = Union[List, Tuple, ArrayLike, "Series", MutableMapping]
```
Comment (jreback) [inline pandas/core/tools/datetimes.py:63]: hmm, i would rather add this specifically where its needed as we often times don't actually accept a MutableMapping (yes we *could* but we have specific checks and this doesn't pass)

File: pandas/io/date_converters.py [hunk 2]
```diff
@@ -1,35 +1,110 @@
...
+    """
+    warnings.warn(
+        """
+        Use pd.to_datetime(date_col + " " + time_col) instead to get a Pandas Series.
+        Use pd.to_datetime(date_col + " " + time_col).to_pydatetime() instead to get a Numpy array.
+""",  # noqa: E501
+        FutureWarning,
+        stacklevel=2,
+    )
+    return to_datetime(date_col + " " + time_col).to_pydatetime()
```
Comment (jreback) [inline pandas/io/date_converters.py:26]: are these *exactly* the same? `to_datetime` is much more flexible. I would rather not change the actual impl and just deprecate

Comment (jreback) [reply pandas/io/date_converters.py:26]: though if we can show they are the same then this is ok

File: pandas/tests/io/parser/test_parse_dates.py [hunk 3]
```diff
@@ -778,6 +772,7 @@ def test_multi_index_parse_dates(all_parsers, index_col):
     tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.filterwarnings("ignore::FutureWarning")
```
Comment (jreback) [inline pandas/tests/io/parser/test_parse_dates.py:775]: I wouldn't ignore warnings, you can *assert* the warnings

File: pandas/tests/io/parser/test_parse_dates.py [hunk 4]
```diff
@@ -1263,7 +1255,7 @@ def test_parse_date_time_multi_level_column_name(all_parsers):
 )
 def test_parse_date_time(all_parsers, data, kwargs, expected):
     parser = all_parsers
-    result = parser.read_csv(StringIO(data), date_parser=conv.parse_date_time, **kwargs)
+    result = parser.read_csv(StringIO(data), **kwargs)
```
Comment (jreback) [inline pandas/tests/io/parser/test_parse_dates.py:1258]: i would leave the parsing and assert the warnings, we want to assert that the results are the same as before

File: pandas/tests/io/parser/test_parse_dates.py [hunk 5]
```diff
@@ -1275,12 +1267,7 @@ def test_parse_date_time(all_parsers, data, kwargs, expected):
 def test_parse_date_fields(all_parsers):
     parser = all_parsers
     data = "year,month,day,a\n2001,01,10,10.\n2001,02,1,11."
-    result = parser.read_csv(
-        StringIO(data),
-        header=0,
-        parse_dates={"ymd": [0, 1, 2]},
-        date_parser=conv.parse_date_fields,
-    )
+    result = parser.read_csv(StringIO(data), header=0, parse_dates={"ymd": [0, 1, 2]},)
```
Comment (jreback) [inline pandas/tests/io/parser/test_parse_dates.py:1270]: same comment

**Review by jreback — CHANGES_REQUESTED** [review]
so you are deprecating, changing the impl and removing the tests. I am ok with this if you don't remove any tests (and just assert the warnings). otherwise its very hard to tell if things still work as before (which defeats the point of deprecation). i am also ok with leaving the impl completely (though your changes do clean up).

File: pandas/io/date_converters.py (hunk 2 above)
Comment (avinashpancham) [reply pandas/io/date_converters.py:26]: What do you mean with just deprecating the impl, that we keep the original impl code and only add the warning at the beginning? I would be fine with that.

File: pandas/tests/io/parser/test_parse_dates.py (hunk 4 above)
Comment (avinashpancham) [reply pandas/tests/io/parser/test_parse_dates.py:1258]: Would it be fine to replace date_parser=conv.parse_date_time with date_parser=pd.to_datetime. All the tests still pass and no warning is raised then. Plus it safes us some future changes when the functions in date_converters.py are removed from the codebase.

File: pandas/tests/io/parser/test_parse_dates.py (hunk 5 above)
Comment (avinashpancham) [reply pandas/tests/io/parser/test_parse_dates.py:1270]: Here, I would also like to replace `date_parser=conv.parse_date_fields` with `date_parser=pd.to_datetime`, due to the same reasoning as above

File: pandas/tests/io/parser/test_parse_dates.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/io/parser/test_parse_dates.py:1258]: so we would need to have a test that faithully replicates `conv.parse_date_time` (or hard codes the result). I am pretty sure that `pd.to_datetime` is fully replicative but we need to have tests that prove that.

File: pandas/tests/io/parser/test_parse_dates.py (hunk 5 above)
Comment (jreback) [reply pandas/tests/io/parser/test_parse_dates.py:1270]: sure, same reasoning on tests as above.

its a tossup here on whether to change the impl and deprecate or just deprecate. I agree that we *do* want to change the implementation, but don't also want to break things that we maybe lightly testing.

File: pandas/io/date_converters.py [hunk 6]
```diff
@@ -1,35 +1,110 @@
...
     )
 
+    return np.array([s.to_pydatetime() for s in ser])
+
 
 def generic_parser(parse_func, *cols):
+    """
+    Use dateparser to parse columns with data information into a single datetime column.
+
+    .. deprecated:: 1.1.0
```
Comment (jreback) [inline pandas/io/date_converters.py:96]: these all should be 1.2 

its ok to just have the warning text, no need to double in the doc-string.

Comment (avinashpancham) [reply pandas/io/date_converters.py]: Done

File: pandas/tests/io/parser/test_parse_dates.py (hunk 3 above)
Comment (avinashpancham) [reply pandas/tests/io/parser/test_parse_dates.py]: Done

File: pandas/io/date_converters.py (hunk 2 above)
Comment (avinashpancham) [reply pandas/io/date_converters.py]: I reverted back to the original impl. Now we have the original impl and a warning

File: pandas/core/tools/datetimes.py (hunk 1 above)
Comment (avinashpancham) [reply pandas/core/tools/datetimes.py]: Reverted back to original impl, so MutableMapping was no longer necessary

File: pandas/io/date_converters.py [hunk 7]
```diff
@@ -30,6 +78,20 @@ def parse_all_fields(year_col, month_col, day_col, hour_col, minute_col, second_
...
+    """
+    Use dateparser to parse columns with data information into a single datetime column.
+
+    .. deprecated:: 1.2
+    """
+
+    warnings.warn(
+        """
+        Use pd.to_datetime instead.
+""",  # noqa: E501
```
Comment (MarcoGorelli) [inline pandas/io/date_converters.py:90]: is `noqa` necessary here?

Comment (avinashpancham) [reply pandas/io/date_converters.py]: Not necessary, removed it.

File: pandas/tests/io/test_date_converters.py [hunk 8]
```diff
@@ -8,22 +8,24 @@
...
+
     dates = np.array(["2007/1/3", "2008/2/4"], dtype=object)
     times = np.array(["05:07:09", "06:08:00"], dtype=object)
     expected = np.array([datetime(2007, 1, 3, 5, 7, 9), datetime(2008, 2, 4, 6, 8, 0)])
-
-    result = conv.parse_date_time(dates, times)
-    tm.assert_numpy_array_equal(result, expected)
+    with tm.assert_produces_warning(FutureWarning):
+        result = conv.parse_date_time(dates, times)
+        tm.assert_numpy_array_equal(result, expected)
```
Comment (MarcoGorelli) [inline pandas/tests/io/test_date_converters.py:17]: For this and other tests below, can you just assert than the first line produces a warning?
something like
```
    with tm.assert_produces_warning(FutureWarning):
        result = conv.parse_date_time(dates, times)
    tm.assert_numpy_array_equal(result, expected)
```

Comment (avinashpancham) [reply pandas/tests/io/test_date_converters.py]: Done

File: pandas/tests/io/parser/test_parse_dates.py (hunk 4 above)
Comment (avinashpancham) [reply pandas/tests/io/parser/test_parse_dates.py]: Added a pytest dateparser parameter that tests both the old function and the new function (`pd.to_datetime`). Both functions produce the same expected value, so I think we show that the new function is replicative.

**Comment by avinashpancham** [comment]
@jreback addressed all your comments. Could you have a second look?

**Comment by jreback** [comment]
this looks good can you merge master and ping on green.

**Comment by avinashpancham** [comment]
@jreback merged master; all checks in CI pass except two due to a flaky test (see below)

`hypothesis.errors.Flaky: Hypothesis test_shift_across_dst(offset=<31 * MonthEnds>) produces unreliable results: Falsified on the first call but did not on a subsequent one`

**Comment by avinashpancham** [comment]
Reran CI, now everything passes

**Review by jreback — APPROVED** [review]

File: doc/source/whatsnew/v1.2.0.rst [hunk 9]
```diff
@@ -195,7 +195,7 @@ Deprecations
 ~~~~~~~~~~~~
 - Deprecated parameter ``inplace`` in :meth:`MultiIndex.set_codes` and :meth:`MultiIndex.set_levels` (:issue:`35626`)
 - Deprecated parameter ``dtype`` in :~meth:`Index.copy` on method all index classes. Use the :meth:`Index.astype` method instead for changing dtype(:issue:`35853`)
--
+- Date parser functions :func:`parse_date_time`, :func:`parse_date_fields`, :func:`parse_all_fields` and :func:`generic_parser` from ``pandas.io.date_converters`` are deprecated and will be removed in a future version; use pandas.to_datetime() instead (GH35741)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:198]: can you put in ```:func:`to_datetime` ``` so it renders all of the others won't render at all i think (e.g. ```:func:`parse_date_time` ```. so those should be updated to a fully qualified path (e.g. pandas.io.date_converters)

**Review by jreback — CHANGES_REQUESTED** [review]
looks good some doc comments, also please merge master. ping on green.

**Comment by avinashpancham** [comment]
@jreback updated doc, merged master and CI on green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @avinashpancham very nice!

**Pull request merged**